src/qcodes/instrument_drivers/cryomagnetics/Cryomagnetics4G.py

```diff
@@ -0,0 +1,376 @@
+        ramping_state_check_interval (Numbers): The interval in seconds to check the ramping state.
+        field (Numbers): The magnetic field strength in Tesla.
+        rate (Numbers): The ramp rate for the magnetic field in Tesla per minute.
+        Vmag (Numbers): The magnet sense voltage in Volts.
+        Vout (Numbers): The magnet output voltage in Volts.
+        Iout (Numbers): The magnet output field/current in Amps.
+
+    Functions:
+        QReset: Resets the quench condition.
+        remote: Sets the instrument to remote mode.
+        off: Ramps the magnetic field to zero (non-blocking).
+    """
+    KG_TO_TESLA = 0.1  # Constant for unit conversion
+
+    def __init__(
+        self,
+        name: str,
+        address: str,
+        max_current_limits: dict[int, tuple[float, float]],
+        coil_constant: float,
+        **kwargs,
+    ):
+        super().__init__(name, address, terminator="\n", **kwargs)
+
+        self.coil_constant = coil_constant
+        self.max_current_limits = max_current_limits
+
+        # Initialize  rate manager based on hypothetical hardware specific limits
+        self._initialize_max_current_limits()
+
+        # Adding parameters
+        self.add_parameter(
+            name="units",
+            set_cmd="UNITS {}",
+            get_cmd="UNITS?",
+            get_parser=str,
+            vals=Enum("A", "kG", "T"),
+            docstring="Field Units",
+        )
+
+        self.add_parameter(
+            "ramping_state_check_interval",
+            initial_value=0.05,
+            unit="s",
+            vals=Numbers(0, 10),
+            set_cmd=None,
+        )
+
+        self.add_parameter(
+            name="field",
+            unit="T",
+            set_cmd=self.set_field,
+            get_cmd=self._get_field,
+            get_parser=float,
+            vals=Numbers(-9.001, 9.001),
+            docstring="Magnetic Field in Tesla",
+        )
+
+        self.add_parameter(
+            name="rate",
+            unit="T/min",
+            get_cmd=self._get_rate,
+            set_cmd=self._set_rate,
+            get_parser=float,
+            docstring="Rate for magnetic field T/min",
+        )
+
+        self.add_parameter(
+            name="Vmag",
+            unit="V",
+            get_cmd="VMAG?",
+            get_parser=float,
+            vals=Numbers(-10, 10),
+            docstring="Magnet sense voltage",
+        )
+
+        self.add_parameter(
+            name="Vout",
+            unit="V",
+            get_cmd="VOUT?",
+            get_parser=float,
+            vals=Numbers(-12.8, 12.8),
+            docstring="Magnet output voltage",
+        )
+
+        self.add_parameter(
+            name="Iout",
+            unit="A",
+            get_cmd="IOUT?",
+            get_parser=float,
+            docstring="Magnet output field/current",
+        )
+
+        # Add function to reset quench
+        self.add_function("QReset", call_cmd="QRESET")
```

Comment:
We don't recommend using functions any more since they add no value over having a method.